### PR TITLE
ApiTests. remove objects from space.

### PR DESCRIPTION
**Pull request by ScharfViktor**
- added test "create subfolder"
- added the ability to search objects in subfolders of the space 
- added tests "deleting object from space" and "deleting space"
- created bug: https://github.com/owncloud/ocis/issues/3031

### CODE REVIEW

File: tests/acceptance/features/apiSpaces/uploadSpaces.feature [hunk 1]
```diff
@@ -11,20 +11,32 @@ Feature: Upload files into a space
...
+
   Scenario: A user can create a folder in a Space via the Graph API
     Given user "Alice" has created a space "Project Ceres" of type "project" with quota "2000"
     When user "Alice" creates a folder "mainFolder" in space "Project Ceres" using the WebDav Api
     Then the HTTP status code should be "201"
     And for user "Alice" the space "Project Ceres" should contain these entries:
       | mainFolder        |
 
+
+  Scenario: A user can create a subfolders in a Space via the Graph API
```
Comment (phil-davis) [inline tests/acceptance/features/apiSpaces/uploadSpaces.feature:23]:
```suggestion
  Scenario: A user can create subfolders in a Space via the Graph API
```

File: tests/acceptance/features/apiSpaces/uploadSpaces.feature [hunk 2]
```diff
@@ -11,20 +11,32 @@ Feature: Upload files into a space
...
+    Given user "Alice" has created a space "Subfolders in Project" of type "project" with quota "2000"
+    When user "Alice" creates a subfolder "mainFolder/subFolder1/subFolder2" in space "Subfolders in Project" using the WebDav Api
+    Then the HTTP status code should be "201"
+    And for user "Alice" the space "Subfolders in Project" should contain these entries:
+      | mainFolder |
+    And for user "Alice" folder "mainFolder/subFolder1/" of the space "Subfolders in Project" should contain these entries:
+      | subFolder2 |
+
```
Comment (phil-davis) [inline tests/acceptance/features/apiSpaces/uploadSpaces.feature:31]: You could add a 2nd blank line here, to be the same as the other scenarios.

File: tests/acceptance/features/apiSpaces/removeSpaces.feature [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    And for user "Brian" the space "Editor deletes file" should not contain these entries:
+      | text.txt |
+    When user "Brian" lists all available spaces via the GraphApi
+    Then the json responded should contain a space "Editor deletes file" with these key and value pairs:
+      | key          | value              |
+      | name         | Editor deletes file |
+      | quota@@@used | 0                  |
+
+
+  Scenario: A editor can delete a empty Space via the webDav API
```
Comment (phil-davis) [inline tests/acceptance/features/apiSpaces/removeSpaces.feature:108]: I am a bit surprised by this. But maybe this is the requirement?
I was thinking that spaces would have at least 3 roles like:
- Viewer - read-only of everything in the space
- Editor - can create/change/delete the resources (files and folders) in the space
- Manager - can manage access to the space (grant and revoke roles to users...) and can delete the whole space.

File: tests/acceptance/features/apiSpaces/removeSpaces.feature [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    Given user "Alice" has created a space "Viewer deletes file" of type "project" with quota "20"
+    And user "Alice" has uploaded a file inside space "Viewer deletes file" with content "some content" to "text.txt"
+    And user "Alice" has shared a space "Viewer deletes file" to user "Brian" with role "viewer"
+    When user "Brian" removes the object "text.txt" from space "Viewer deletes file"
+    Then the HTTP status code should be "403"
+    And for user "Brian" the space "Viewer deletes file" should contain these entries:
+      | text.txt |
+
+
+  Scenario: A viewer cannot delete a empty Space via the webDav API
```
Comment (phil-davis) [inline tests/acceptance/features/apiSpaces/removeSpaces.feature:160]:
```suggestion
  Scenario: A viewer cannot delete an empty Space via the webDav API
```

File: tests/acceptance/features/apiSpaces/removeSpaces.feature [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+  Scenario: A editor can delete a empty Space via the webDav API
+    Given user "Alice" has created a space "Editor deletes empty space" of type "project" with quota "20"
+    And user "Alice" has shared a space "Editor deletes empty space" to user "Brian" with role "editor"
+    When user "Brian" removes a space "Editor deletes empty space"
+    Then the HTTP status code should be "200"
+    When user "Brian" lists all available spaces via the GraphApi
+    Then the json responded should not contain a space with name "Editor deletes empty space"
+
+
+  Scenario: A editor can delete a Space containing folders and files via the webDav API
```
Comment (phil-davis) [inline tests/acceptance/features/apiSpaces/removeSpaces.feature:117]:
```suggestion
  Scenario: An editor can delete a Space containing folders and files via the webDav API
```

File: tests/acceptance/features/apiSpaces/removeSpaces.feature (hunk 3 above)
Comment (phil-davis) [inline tests/acceptance/features/apiSpaces/removeSpaces.feature:108]:
```suggestion
  Scenario: An editor can delete an empty Space via the webDav API
```

File: tests/acceptance/features/apiSpaces/removeSpaces.feature [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+    And user "Alice" has shared a space "Editor deletes subfolder" to user "Brian" with role "editor"
+    When user "Brian" removes the object "folder/subFolderForDeleting" from space "Editor deletes subfolder"
+    Then the HTTP status code should be "200"
+    And for user "Brian" the space "Editor deletes subfolder" should contain these entries:
+      | folder |
+    And for user "Brian" folder "folder/" of the space "Editor deletes subfolder" should not contain these entries:
+      | subFolderForDeleting |
+
+
+  Scenario: A editor can delete a file in a Space via the webDav API
```
Comment (phil-davis) [inline tests/acceptance/features/apiSpaces/removeSpaces.feature:93]:
```suggestion
  Scenario: An editor can delete a file in a Space via the webDav API
```

File: tests/acceptance/features/apiSpaces/removeSpaces.feature [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+    Given user "Alice" has created a space "Editor deletes folder" of type "project" with quota "10"
+    And user "Alice" has created a folder "folderForDeleting/sub1/sub2" in space "Editor deletes folder"
+    And user "Alice" has shared a space "Editor deletes folder" to user "Brian" with role "editor"
+    When user "Brian" removes the object "folderForDeleting" from space "Editor deletes folder"
+    Then the HTTP status code should be "200"
+    And for user "Brian" the space "Editor deletes folder" should not contain these entries:
+      | folderForDeleting |
+
+
+  Scenario: A editor can delete a subfolder in a Space via the webDav API
```
Comment (phil-davis) [inline tests/acceptance/features/apiSpaces/removeSpaces.feature:81]:
```suggestion
  Scenario: An editor can delete a subfolder in a Space via the webDav API
```

File: tests/acceptance/features/apiSpaces/removeSpaces.feature [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+    And user "Alice" has created a folder "folderForDeleting/sub1/sub2" in space "Owner deletes space"
+    And user "Alice" has uploaded a file inside space "Owner deletes space" with content "some content" to "folderForDeleting/sub1/sub2/text.txt"
+    When user "Alice" removes a space "Owner deletes space"
+    Then the HTTP status code should be "200"
+    When user "Alice" lists all available spaces via the GraphApi
+    Then the json responded should not contain a space with name "Owner deletes space"
+
+#   editor role
+
+  Scenario: A editor can delete a folder with some subfolders in a Space via the webDav API
```
Comment (phil-davis) [inline tests/acceptance/features/apiSpaces/removeSpaces.feature:71]:
```suggestion
  Scenario: An editor can delete a folder with some subfolders in a Space via the webDav API
```

File: tests/acceptance/features/apiSpaces/removeSpaces.feature [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+
+  Scenario: A owner can delete a empty Space via the webDav API
+    Given user "Alice" has created a space "Owner deletes empty space" of type "project" with quota "20"
+    When user "Alice" removes a space "Owner deletes empty space"
+    Then the HTTP status code should be "200"
+    When user "Alice" lists all available spaces via the GraphApi
+    Then the json responded should not contain a space with name "Owner deletes empty space"
+
+
+  Scenario: A owner can delete a Space containing folders and files via the webDav API
```
Comment (phil-davis) [inline tests/acceptance/features/apiSpaces/removeSpaces.feature:60]:
```suggestion
  Scenario: An owner can delete a Space containing folders and files via the webDav API
```

File: tests/acceptance/features/apiSpaces/removeSpaces.feature [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+    And for user "Alice" the space "Owner deletes file" should not contain these entries:
+      | text.txt |
+    When user "Alice" lists all available spaces via the GraphApi
+    Then the json responded should contain a space "Owner deletes file" with these key and value pairs:
+      | key          | value             |
+      | name         | Owner deletes file |
+      | quota@@@used | 0                 |
+
+
+  Scenario: A owner can delete a empty Space via the webDav API
```
Comment (phil-davis) [inline tests/acceptance/features/apiSpaces/removeSpaces.feature:52]:
```suggestion
  Scenario: An owner can delete an empty Space via the webDav API
```

File: tests/acceptance/features/apiSpaces/removeSpaces.feature [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+    And user "Alice" has created a subfolder "folder/subFolderForDeleting" in space "Owner deletes subfolder"
+    When user "Alice" removes the object "folder/subFolderForDeleting" from space "Owner deletes subfolder"
+    Then the HTTP status code should be "200"
+    And for user "Alice" the space "Owner deletes subfolder" should contain these entries:
+      | folder |
+    And for user "Alice" folder "folder/" of the space "Owner deletes subfolder" should not contain these entries:
+      | subFolderForDeleting |
+
+
+  Scenario: A owner can delete a file in a Space via the webDav API
```
Comment (phil-davis) [inline tests/acceptance/features/apiSpaces/removeSpaces.feature:38]:
```suggestion
  Scenario: An owner can delete a file in a Space via the webDav API
```

File: tests/acceptance/features/apiSpaces/removeSpaces.feature [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+  Scenario: A owner can delete a folder with some subfolders in a Space via the webDav API
+    Given user "Alice" has created a space "Owner deletes folder" of type "project" with quota "10"
+    And user "Alice" has created a folder "folderForDeleting/sub1/sub2" in space "Owner deletes folder"
+    When user "Alice" removes the object "folderForDeleting" from space "Owner deletes folder"
+    Then the HTTP status code should be "200"
+    And for user "Alice" the space "Owner deletes folder" should not contain these entries:
+      | folderForDeleting |
+
+
+  Scenario: A owner can delete a subfolder in a Space via the webDav API
```
Comment (phil-davis) [inline tests/acceptance/features/apiSpaces/removeSpaces.feature:27]:
```suggestion
  Scenario: An owner can delete a subfolder in a Space via the webDav API
```

File: tests/acceptance/features/apiSpaces/removeSpaces.feature [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+  See https://github.com/owncloud/ocis/issues/1542 and https://github.com/owncloud/ocis/pull/839
+
+  Background:
+    Given user "Alice" has been created with default attributes and without skeleton files
+    And user "Brian" has been created with default attributes and without skeleton files
+    And the administrator has given "Alice" the role "Admin" using the settings api
+
+#   owner of space
+
+  Scenario: A owner can delete a folder with some subfolders in a Space via the webDav API
```
Comment (phil-davis) [inline tests/acceptance/features/apiSpaces/removeSpaces.feature:18]:
```suggestion
  Scenario: An owner can delete a folder with some subfolders in a Space via the webDav API
```

**Review by phil-davis — COMMENTED** [review]
LGTM - mostly just the English word "a" and "an" :)

File: tests/acceptance/features/apiSpaces/removeSpaces.feature (hunk 3 above)
Comment (ScharfViktor) [reply tests/acceptance/features/apiSpaces/removeSpaces.feature:108]: AC in https://jira.owncloud.com/browse/OCIS-2063
viewer has no right to delete

Comment (ScharfViktor) [reply tests/acceptance/features/apiSpaces/removeSpaces.feature]: You are right, the editor should not be able to delete space.
https://jira.owncloud.com/browse/OCIS-2043
I will collect more information tomorrow and change some cases

File: tests/acceptance/features/apiSpaces/removeSpaces.feature [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+
+  Note - this feature is run in CI with ACCOUNTS_HASH_DIFFICULTY set to the default for production
+  See https://github.com/owncloud/ocis/issues/1542 and https://github.com/owncloud/ocis/pull/839
+
+  Background:
+    Given user "Alice" has been created with default attributes and without skeleton files
+    And user "Brian" has been created with default attributes and without skeleton files
+    And the administrator has given "Alice" the role "Admin" using the settings api
+
+#   owner of space
```
Comment (individual-it) [inline tests/acceptance/features/apiSpaces/removeSpaces.feature:16]: what is this line about?

**Comment by ScharfViktor** [comment]
Paused. I am waiting for permissions change from @C0rby.
for disabling/removing space we should use another method: `curl -k -s -u admin:admin -X DELETE 'https://localhost:9200/graph/v1.0/drives/space_uuid'`
Tests will be changed

**Comment by C0rby** [comment]
The current ocis master now contain the permission changes.
Disabling and deleting a space should now work in ocis master.

File: tests/acceptance/features/apiSpaces/removeSpaces.feature (hunk 14 above)
Comment (ScharfViktor) [reply tests/acceptance/features/apiSpaces/removeSpaces.feature]: I separated tests by permissions: owner, editor, viewer

**Comment by ScharfViktor** [comment]
I changed tests and bug report https://github.com/owncloud/ocis/issues/3031. PR is ready for review

I have added 3 tests for expected failures. In the bug report I described what was wrong in my opinion

**Review by individual-it — APPROVED** [review]
Tests look good to me, but we should discuss if adding the WebDAV tests is necessary, currently there is a lot of work going on to add "spaces" as a variant to the existing WebDAV tests. That should cover most WebDAV cases I think
see https://github.com/owncloud/ocis/issues/2854

**Pull request merged**